### PR TITLE
MetadataStore: Upgrade tool

### PR DESCRIPTION
**Pull request by dushyanthsc**
Fixes https://github.com/kubeflow/pipelines/issues/3284

As per this [published contract](https://docs.google.com/document/d/1gF-mx3lMyU9h7MAAOXP-KGV-BF-UabDsAlFrWNNhKBo/edit?usp=sharing) shared with the Kubeflow Pipelines community this tool provides Metadata Store upgrade support in KFP clusters.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/3295)

### CODE REVIEW

**Comment by rmgogogo** [comment]
/lgtm

Would you ping joeliedtke to request a doc update in kubeflow/website project?

**Comment by Bobgy** [comment]
Great work with the tool! Looks configurable and easy to use.

I've got some general questions, probably we can document it here too:
* > The upgrade proposal does not cover upgrading any existing clusters, but refers to clusters created after the feature is complete. For existing clusters, the workflow and tool developed as part of the proposal can be used but without guarantees.

    From which version do we support upgrade?
* How to know current version?
* Is there a user facing documentation that includes just the information in the [doc](https://docs.google.com/document/d/1gF-mx3lMyU9h7MAAOXP-KGV-BF-UabDsAlFrWNNhKBo/edit#) that a user needs?
* Can we distribute executable binary too? I'm not sure every user is familiar with setting up go envs

Probably as a TODO item:
We should have a dedicated page in kubeflow.org linked from the upgrade section: https://www.kubeflow.org/docs/pipelines/upgrade/.

File: tools/metadatastore-upgrade/main.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+)
+
+var (
+	kubeconfigPath = flag.String("kubeconfig", "", "Absolute path to kubeconfig file.")
+	imageTag = flag.String("image_tag", "", "Image tag to use for upgrade")
+	deploymentNamespace = flag.String("namespace", "kubeflow", "Namespace where MetadataStore is deployed.")
+)
+
+const (
+	ML_METADATA_DEPLOYMENT = "metadata-deployment"
```
Comment (elikatsis) [inline tools/metadatastore-upgrade/main.go]: How about changing this to `metadata-grpc-deployment` to be aligned with kubeflow/pipelines#2889  ?

File: tools/metadatastore-upgrade/main.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	originalImage := oldDeployment.Spec.Template.Spec.Containers[0].Image
+	originalContainerArgs := oldDeployment.Spec.Template.Spec.Containers[0].Args
+
+	newImage := ML_METADATA_IMAGE + ":" + *imageTag
+	log.Printf("Upgrading MetadataStore in Namespace: %s to Image: %s", *deploymentNamespace, newImage)
+	updateErr := UpdateDeployment(deploymentsClient, newImage, append(originalContainerArgs, UPGRADE_FLAG))
+	if updateErr == nil {
+	  log.Printf("MetadataStore successfully upgraded to Image: %s", newImage)
+	  log.Printf("Cleaning up Upgrade")
+		upgradeCleanupErr := 	UpdateDeployment(deploymentsClient, newImage, originalContainerArgs)
```
Comment (neuromage) [inline tools/metadatastore-upgrade/main.go]: If we left it as is, what is the downside?

/cc @hughmiao

File: tools/metadatastore-upgrade/main.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+		// In a highly unlikely scenario upgrade cleanup can fail.
+		if upgradeCleanupErr != nil {
+			log.Printf("Upgrade cleanup failed. Likely MetadataStore is in a functioning state. Please contact Kubeflow community for verification")
+			os.Exit(1)
+		}
+	} else {
+		log.Printf("MetadataStore Upgrade failed. Reverting back to Old Image")
+		revertUpdateErr := UpdateDeployment(deploymentsClient, originalImage, originalContainerArgs)
+		if revertUpdateErr != nil {
+			log.Printf("Revert failed. Please contact Kubeflow community")
```
Comment (neuromage) [inline tools/metadatastore-upgrade/main.go]: I don't think the error should ask users to contact KF community. 

Also, we probably want to print the error when it happens

File: tools/metadatastore-upgrade/main.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+			os.Exit(1)
+		} else {
+			log.Printf("Revert succeeded. MetadataStore unchanged.")
+			os.Exit(1)
```
Comment (neuromage) [inline tools/metadatastore-upgrade/main.go]: Is this os.Exit needed here?

**Comment by dushyanthsc** [comment]
Answers Inline

> Great work with the tool! Looks configurable and easy to use.
> 
> I've got some general questions, probably we can document it here too:
> 
> * > The upgrade proposal does not cover upgrading any existing clusters, but refers to clusters created after the feature is complete. For existing clusters, the workflow and tool developed as part of the proposal can be used but without guarantees.
>   
>   
>   From which version do we support upgrade?
 --- Updated README to provide details.
> * How to know current version?
--- Updated README to provide details.
> * Is there a user facing documentation that includes just the information in the [doc](https://docs.google.com/document/d/1gF-mx3lMyU9h7MAAOXP-KGV-BF-UabDsAlFrWNNhKBo/edit#) that a user needs?
--- Will follow up with joeliedtke.
> * Can we distribute executable binary too? I'm not sure every user is familiar with setting up go envs
 Good Idea. Will follow in a separate issue.
> Probably as a TODO item:
> We should have a dedicated page in kubeflow.org linked from the upgrade section: https://www.kubeflow.org/docs/pipelines/upgrade/.
Will contact joeliedtke for this.

File: tools/metadatastore-upgrade/main.go (hunk 1 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Agreed. Will change this and the documentation once #2889 is merged

File: tools/metadatastore-upgrade/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# MetadataStore Upgrade Tool
+
+As per this [published contract](https://docs.google.com/document/d/1gF-mx3lMyU9h7MAAOXP-KGV-BF-UabDsAlFrWNNhKBo/edit?usp=sharing) shared with the Kubeflow Pipelines community this tool provides Metadata Store upgrade support in KFP clusters.
+
+To run the tool execute the following command from this folder:
+
+```
+go run main.go --image_tag=<image-tag> --kubeconfig=<kubeconfig-path> --namespace=<namespace-name>
+
+ -- image_tag(Requried) - Image tag of a published MetadataStore image in the gcr repository - gcr.io/tfx-oss-public/ml_metadata_store_server
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/README.md]: Done.

File: tools/metadatastore-upgrade/README.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# MetadataStore Upgrade Tool
+
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/README.md]: Added more information to the README file

File: tools/metadatastore-upgrade/main.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+	"k8s.io/client-go/util/homedir"
+	"k8s.io/client-go/util/retry"
+
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+var (
+	kubeconfigPath = flag.String("kubeconfig", "", "Absolute path to kubeconfig file.")
+	imageTag = flag.String("image_tag", "", "Image tag to use for upgrade")
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Changed it to `new_image_tag` and updated the description to clarify.

File: tools/metadatastore-upgrade/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Done.

File: tools/metadatastore-upgrade/main.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+)
+
+const (
+	ML_METADATA_DEPLOYMENT = "metadata-deployment"
+	ML_METADATA_IMAGE = "gcr.io/tfx-oss-public/ml_metadata_store_server"
+	UPGRADE_FLAG = "--enable_database_upgrade=true"
+	MAX_WAIT_TIME = 30
+)
+
+func UpdateDeployment(deploymentsClient v1.DeploymentInterface, image string, containerArgs []string) error {
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Done.

File: tools/metadatastore-upgrade/main.go [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			// Retrieve the latest version of Deployment before attempting update
+			// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+			result, getErr := deploymentsClient.Get( ML_METADATA_DEPLOYMENT,  metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Done.

File: tools/metadatastore-upgrade/main.go [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+			if getErr != nil {
+				return fmt.Errorf("Failed to get latest version of Deployment: %v", getErr)
+			}
+
+			result.Spec.Template.Spec.Containers[0].Image = image
+			result.Spec.Template.Spec.Containers[0].Args = containerArgs
+			_, updateErr := deploymentsClient.Update(result)
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Done.

File: tools/metadatastore-upgrade/main.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+			if updateErr != nil {
+				return updateErr
+			}
+
+			result, getErr = deploymentsClient.Get( ML_METADATA_DEPLOYMENT,  metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Done.

File: tools/metadatastore-upgrade/main.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+	})
+}
+
+
+func main() {
+	flag.Parse()
+	log.SetOutput(os.Stdout)
+
+	if *imageTag == "" {
+		log.Printf("Missing image_tag flag")
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Done.

File: tools/metadatastore-upgrade/main.go [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+	deploymentsClient := clientset.AppsV1().Deployments(*deploymentNamespace)
+
+	oldDeployment, getErr := deploymentsClient.Get( ML_METADATA_DEPLOYMENT,  metav1.GetOptions{})
+	if getErr != nil {
+		log.Printf("Failed to get old Deployment: %v", getErr)
+		os.Exit(1)
+	}
+
+	originalImage := oldDeployment.Spec.Template.Spec.Containers[0].Image
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Done.

File: tools/metadatastore-upgrade/main.go [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+	originalContainerArgs := oldDeployment.Spec.Template.Spec.Containers[0].Args
+
+	newImage := ML_METADATA_IMAGE + ":" + *imageTag
+	log.Printf("Upgrading MetadataStore in Namespace: %s to Image: %s", *deploymentNamespace, newImage)
```
[the remark this answers is not in the document]
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Done.

File: tools/metadatastore-upgrade/main.go (hunk 2 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: The initial fresh install needs to have `--enable_database_upgrade=false`. The problem with not cleaning up the argument after upgrade is that future upgrade runs will keep adding the arguments to the list. We could have some logic to check if the argument is already there but its just too messy.

File: tools/metadatastore-upgrade/main.go (hunk 3 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Logged the error and cleaned up the log statement.

File: tools/metadatastore-upgrade/main.go (hunk 4 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Yes because the reason for running the tool was un-successful. It would be better to return failure in cases where this tool is used as part of a script.

File: tools/metadatastore-upgrade/README.md [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+Upgrade tool provides a mechanism for KFP users to upgrade MetadataStore component in their KFP cluster. A MetadatatStore upgrade is composed of two aspects:
```
Comment (neuromage) [inline tools/metadatastore-upgrade/README.md]: typo: MetadataStore

File: tools/metadatastore-upgrade/README.md [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+* Upgrading the image used in `metadata-deployment` K8s Deployment. The user is expected to provide an image-tag for this upgrade.
```
Comment (neuromage) [inline tools/metadatastore-upgrade/README.md]: Good to provide a link here to the canonical definition? Maybe standalone manifests?

File: tools/metadatastore-upgrade/README.md [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+* Upgrading the MYSQL database schema to adhere to MLMD library used in the `metadata-deployment` image. The tool automatically handles schema upgrade using a [feature](https://github.com/google/ml-metadata/releases/tag/v0.21.0) in the gRPC server used in the `metadata-deployment` image.
```
Comment (neuromage) [inline tools/metadatastore-upgrade/README.md]: Do you mean to say the gRPC server handles the schema migration of the backing mySQL database if some flag is set to true? If so, maybe reword this sentence to better reflect that, and link to the flag definition?

File: tools/metadatastore-upgrade/README.md [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+The contract for this tool was published and shared with Kubeflow Pipelines community in this [doc](https://docs.google.com/document/d/1gF-mx3lMyU9h7MAAOXP-KGV-BF-UabDsAlFrWNNhKBo/edit?usp=sharing)
+
+To run the tool execute the following command from this folder:
+
+```
+go run main.go --new_image_tag=<image-tag> --kubeconfig=<kubeconfig-path> --namespace=<namespace-name>
+```
+
+Arguments:
+* `--new_image_tag`(Required) - Tag of a published ML-Metadata Store Server image in this [gcr repository](gcr.io/tfx-oss-public/ml_metadata_store_server)
```
Comment (neuromage) [inline tools/metadatastore-upgrade/README.md]: "The image tag for the gRPC server version to upgrade to. The list of available images can be found here..."

File: tools/metadatastore-upgrade/README.md [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+* `--kubeconfig`(Optional) - Absolute path to a kubeconfig file. If this argument is not specified `.kubecofing` in user's home directory is used.
+* `--namespace`(Optional) - Namespace where `metadata-deployment` is deployed in the KFP cluster. Defaults to `kubeflow`.
+
+**Note:** 
+1. This upgrade tool should be the only client interacting with the MetadataStore during upgrade.
```
Comment (neuromage) [inline tools/metadatastore-upgrade/README.md]: Call this out more prominently at the top of the file?

File: tools/metadatastore-upgrade/main.go [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+	newImage := mlMetadataImage + ":" + *imageTag
+	log.Printf("Upgrading MetadataStore in Namespace: %s from Image: %s to Image: %s", *deploymentNamespace, originalImage, newImage)
+	err = updateDeployment(deploymentsClient, newImage, append(originalContainerArgs, upgradeFlag))
+	if err == nil {
+		log.Printf("MetadataStore successfully upgraded to Image: %s", newImage)
+		log.Printf("Cleaning up Upgrade")
+		err = updateDeployment(deploymentsClient, newImage, originalContainerArgs)
+		// In a highly unlikely scenario upgrade cleanup can fail.
+		if err != nil {
+			log.Fatalf("Upgrade cleanup failed: %v. \nLikely MetadataStore is in a functioning state.", err)
```
Comment (neuromage) [inline tools/metadatastore-upgrade/main.go]: At this point, should the user manually attempt to edit the deployment to remove the flag? Also, should we just print this error, and exit successfully, since the upgrade part worked?

File: tools/metadatastore-upgrade/main.go [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+	originalImage := originalDeployment.Spec.Template.Spec.Containers[0].Image
+	originalContainerArgs := originalDeployment.Spec.Template.Spec.Containers[0].Args
+
+	newImage := mlMetadataImage + ":" + *imageTag
+	log.Printf("Upgrading MetadataStore in Namespace: %s from Image: %s to Image: %s", *deploymentNamespace, originalImage, newImage)
+	err = updateDeployment(deploymentsClient, newImage, append(originalContainerArgs, upgradeFlag))
+	if err == nil {
+		log.Printf("MetadataStore successfully upgraded to Image: %s", newImage)
+		log.Printf("Cleaning up Upgrade")
+		err = updateDeployment(deploymentsClient, newImage, originalContainerArgs)
```
Comment (neuromage) [inline tools/metadatastore-upgrade/main.go]:
```
if err := ....; err != nil {
...
```

File: tools/metadatastore-upgrade/main.go [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+		// In a highly unlikely scenario upgrade cleanup can fail.
+		if err != nil {
+			log.Fatalf("Upgrade cleanup failed: %v. \nLikely MetadataStore is in a functioning state.", err)
+		}
+	} else {
+		log.Printf("MetadataStore Upgrade failed: %v. \nReverting back to Old Image", err)
```
Comment (neuromage) [inline tools/metadatastore-upgrade/main.go]: What kind of failure can happen in line 128? Should we just fail, rather than attempt to revert here? The old image could still be running intact?

File: tools/metadatastore-upgrade/main.go (hunk 1 above)
Comment (elikatsis) [reply tools/metadatastore-upgrade/main.go]: Ping for this, since the PR got merged :smile:

File: tools/metadatastore-upgrade/README.md (hunk 15 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/README.md]: Fixed.

File: tools/metadatastore-upgrade/README.md (hunk 16 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/README.md]: Done.

File: tools/metadatastore-upgrade/README.md (hunk 17 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/README.md]: Done.

File: tools/metadatastore-upgrade/README.md (hunk 18 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/README.md]: Fixed.

File: tools/metadatastore-upgrade/README.md (hunk 19 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/README.md]: Done.

File: tools/metadatastore-upgrade/main.go (hunk 20 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: I am ok taking that approach. Updated code to reflect that.

File: tools/metadatastore-upgrade/main.go (hunk 21 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Done.

File: tools/metadatastore-upgrade/main.go (hunk 22 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: I am ok taking that approach. Updated the code to reflect that approach.

File: tools/metadatastore-upgrade/main.go (hunk 1 above)
Comment (dushyanthsc) [reply tools/metadatastore-upgrade/main.go]: Done.

File: tools/metadatastore-upgrade/README.md [hunk 23]
```diff
@@ -0,0 +1,35 @@
+# MetadataStore Upgrade Tool
+
```
Comment (rmgogogo) [inline tools/metadatastore-upgrade/README.md:2]: how about add a sentence that user has these options:
(1) upgrade whole KFP system (via CLI guide/doc), so naturally MLMD server will be upgraded (MLMD server handles MLMD DB's migration).
(2) if user don't want to upgrade whole KFP system, then follow this tool to just upgrade MLMD server (MLMD server handles MLMD DB's migration as long as --enable_database_upgrade=true which is suggested).

I may didn't get the reason on why we provide this tool. I thought "kubectl apply" with MLMD new manifest would be easier and more acceptable by k8s admin.

Is it more for the reason that we didn't set "--enable_database_upgrade=true" before and want to help existing user correct it? If so, I think a simple yaml just for MLMD with "kubectl apply" can also help on it.

Comment (rmgogogo) [reply tools/metadatastore-upgrade/README.md:2]: "--enable_database_upgrade=true" 
Yuan mentioned you two synced on this part that if enable this flag for MLMD gRPC server, the MLMD scheme will be dirty? If so, I think it's an issue MLMD team should fix, right?

@dushyanthsc @Bobgy

Comment (dushyanthsc) [reply tools/metadatastore-upgrade/README.md:2]: @rmgogogo The "--enable_database_upgrade=true" is not supported to be set by default. The MLMD has a work item to fix this. In the meantime we provide this tool to the OSS community to be able to run pipelines that have been compiled (outside the cluster) with different version of MLMD.

Comment (dushyanthsc) [reply tools/metadatastore-upgrade/README.md:2]: I have also updated the README with your suggestion.

Comment (rmgogogo) [reply tools/metadatastore-upgrade/README.md:2]: got it, thanks.

https://github.com/google/ml-metadata/blob/master/ml_metadata/metadata_store/metadata_store_server_main.cc#L173

For new installations. If we enable this flag in our YAML, for the new installations, it will be enabled and then user no need to manually use this tool, right? Aka, as long as user install a newer version, all done automatically.

https://github.com/kubeflow/pipelines/search?q=enable_database_upgrade&unscoped_q=enable_database_upgrade
